add apalis-imx8 boot firmware update support and fixup for imx8qm-mek erase in mfgtool

mike-scott opened the pull request.

mike-scott commented:
Testing these changes

mike-scott commented:
This needs an env size patch to avoid overflow error.

Tim-Anderson reviewed (approved):
LGTM

igoropaniuk commented on meta-lmp-bsp/recipes-bsp/u-boot/u-boot-fio/apalis-imx8/lmp-base.cfg:
```diff
@@ -3,8 +3,8 @@ CONFIG_SPL_OF_CONTROL=y
 CONFIG_FIT_SIGNATURE=y
-CONFIG_FIT_VERBOSE=y
 CONFIG_SPL_FIT_SIGNATURE=y
+CONFIG_FIT_VERBOSE=y
```
order change not needed

mike-scott replied:
This was to make the comparison between files match up better.

See commit title: bsp: u-boot-fio: apalis-imx8: sync updates from imx8qm-mek

igoropaniuk commented on meta-lmp-bsp/recipes-bsp/u-boot/u-boot-fio/apalis-imx8/lmp.cfg:
```diff
@@ -12,9 +12,9 @@ CONFIG_SPL_LOAD_FIT=y
-CONFIG_SPL_HASH_SUPPORT=y
 CONFIG_RSA=y
 CONFIG_HASH=y
+CONFIG_SPL_HASH_SUPPORT=y
```
same here

mike-scott replied:
See commit title: bsp: u-boot-fio: apalis-imx8: sync updates from imx8qm-mek

igoropaniuk commented on meta-lmp-bsp/recipes-bsp/u-boot/u-boot-fio/apalis-imx8/lmp-base.cfg:
```diff
@@ -29,5 +29,7 @@ CONFIG_HUSH_PARSER=y
 CONFIG_ENV_FAT_INTERFACE="mmc"
 CONFIG_ENV_FAT_DEVICE_AND_PART="0:1"
+CONFIG_USE_BOOTCOMMAND=y
+CONFIG_BOOTCOMMAND="fatload mmc 0:1 ${scriptaddr} /boot.scr; source ${scriptaddr}"
```
AFAIR `CONFIG_BOOTCOMMAND` should stay the same, as we include boot.itb, and for non-sota builds boot.scr
still stays in the first FAT partition
```
IMAGE_BOOT_FILES_mx8qm = "imx-boot u-boot.itb hdmitxfw.bin ${KERNEL_IMAGETYPE} ${@make_dtb_boot_files(d)} boot.scr uEnv.txt"

```

mike-scott replied:
Dropped this change.

igoropaniuk commented on meta-lmp-bsp/recipes-bsp/u-boot/u-boot-fio/apalis-imx8/lmp.cfg:
```diff
@@ -39,16 +40,22 @@ CONFIG_ENV_IS_IN_FAT=y
 CONFIG_ENV_FAT_DEVICE_AND_PART="0:1"
 CONFIG_USE_BOOTCOMMAND=y
-CONFIG_BOOTCOMMAND="fatload mmc 0:1 ${scriptaddr} /boot.itb; setenv verify 1; source ${scriptaddr}"
+CONFIG_BOOTCOMMAND="setenv verify 1; source ${scriptaddr}"
```
as now both iMX8QM-based targets support pre-loaded boot.itb from U-Boot FIT, also makes sense to drop it completely from `IMAGE_BOOT_FILES_sota_mx8qm` in "lmp-machine-custom.inc":
```
IMAGE_BOOT_FILES_sota_mx8qm = "imx-boot u-boot.itb boot.itb"
```

mike-scott replied:
I'll add this change on top of the current series

igoropaniuk commented on meta-lmp-bsp/recipes-bsp/u-boot/u-boot-ostree-scr-fit/apalis-imx8/boot.cmd:
```diff
@@ -5,28 +5,46 @@ echo "Using freescale_${fdt_file}"
 setenv hdmi_image hdmitxfw.bin
 setenv m4_0_image core0_m4_image.bin
 setenv m4_1_image core1_m4_image.bin
+setenv ramdisk_addr_r 0x8a000000
```
any reason why we redefine it here?
in `include/configs/apalis-imx8.h ` it's there already:
```
 84 #define MEM_LAYOUT_ENV_SETTINGS \                                               
 85         "fdt_addr_r=0x83000000\0" \                                             
 86         "hdp_addr=0x9c000000\0" \                                               
 87         "kernel_addr_r=0x80280000\0" \                                          
 88         "ramdisk_addr_r=0x8a000000\0" \ 
```

mike-scott replied:
This was added when comparing the boot scripts between apalis-imx8 and imx8qm-mek.
I'll drop as it makes for easier comparison but not needed.

mike-scott replied:
Dropped this change

igoropaniuk commented on meta-lmp-bsp/recipes-bsp/u-boot/u-boot-base-scr/apalis-imx8/boot.cmd:
```diff
@@ -1,3 +1,14 @@
+setenv devnum 0
+setenv boot_file Image
+setenv bootfile ${boot_file}
+setenv kernel_addr_r 0x80280000
```
any reason to re-define kernel_addr_r, if it's already in `include/configs/apalis-imx8.h`?

mike-scott replied:
Let me check some if these synchronization changes.

This probably isn't needed.

mike-scott replied:
Dropped the addition of the top of this file.  Distro boot is enabled and this all should work out of the box.

igoropaniuk commented on meta-lmp-bsp/recipes-bsp/u-boot/u-boot-ostree-scr-fit/apalis-imx8/boot.cmd:
```diff
@@ -5,28 +5,46 @@ echo "Using freescale_${fdt_file}"
 setenv m4_1_image core1_m4_image.bin
+setenv ramdisk_addr_r 0x8a000000
+# enable relocation of ramdisk
+setenv initrd_high
```
IIRC enabling relocation caused some side-effects on Apalis iMX8 (I haven't debugged that problem though). Haven't you seen any Linux boot issues after?

mike-scott replied:
Dropped the re-define of ramdisk_addr_r and the relocation of ramdisk

mike-scott commented:
Updated with comments from @igoropaniuk 

Diff from previous version:
```
diff --git a/meta-lmp-bsp/conf/machine/include/lmp-machine-custom.inc b/meta-lmp-bsp/conf/machine/include/lmp-machine-custom.inc
index b63cf4c4..1f9c03a0 100644
--- a/meta-lmp-bsp/conf/machine/include/lmp-machine-custom.inc
+++ b/meta-lmp-bsp/conf/machine/include/lmp-machine-custom.inc
@@ -261,7 +261,7 @@ SOTA_EXTRA_CLIENT_FEATURES_append_sota_apalis-imx6-sec = " fiovb"
 # iMX8QM
 SOTA_CLIENT_FEATURES_append_mx8qm = " ubootenv"
 IMAGE_BOOT_FILES_mx8qm = "imx-boot u-boot.itb hdmitxfw.bin ${KERNEL_IMAGETYPE} ${@make_dtb_boot_files(d)} boot.scr uEnv.txt"
-IMAGE_BOOT_FILES_sota_mx8qm = "imx-boot u-boot.itb boot.itb"
+IMAGE_BOOT_FILES_sota_mx8qm = "imx-boot u-boot.itb"
 LMP_BOOT_FIRMWARE_FILES_mx8qm = "imx-boot u-boot.itb"
 BOOTSCR_LOAD_ADDR_mx8qm = "0x83100000"
 KERNEL_IMAGETYPE_sota_mx8qm = "fitImage"
diff --git a/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-base-scr/apalis-imx8/boot.cmd b/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-base-scr/apalis-imx8/boot.cmd
index 40fb22c8..5a2abb23 100644
--- a/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-base-scr/apalis-imx8/boot.cmd
+++ b/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-base-scr/apalis-imx8/boot.cmd
@@ -1,14 +1,3 @@
-setenv fdt_file ${fdtfile}
-echo "Using freescale_${fdt_file}"
-
-# Default boot type and device
-setenv devtype mmc
-setenv devnum 0
-setenv boot_file Image
-setenv bootfile ${boot_file}
-setenv kernel_addr_r 0x80280000
-setenv mmcroot /dev/mmcblk0p2
-
 fatload ${devtype} ${devnum}:1 ${loadaddr} /uEnv.txt
 env import -t ${loadaddr} ${filesize}
 run bootcmd
diff --git a/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-fio/apalis-imx8/lmp-base.cfg b/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-fio/apalis-imx8/lmp-base.cfg
index 84ce9b16..2211f818 100644
--- a/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-fio/apalis-imx8/lmp-base.cfg
+++ b/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-fio/apalis-imx8/lmp-base.cfg
@@ -29,7 +29,5 @@ CONFIG_HUSH_PARSER=y
 CONFIG_ENV_IS_IN_FAT=y
 CONFIG_ENV_FAT_INTERFACE="mmc"
 CONFIG_ENV_FAT_DEVICE_AND_PART="0:1"
-CONFIG_USE_BOOTCOMMAND=y
-CONFIG_BOOTCOMMAND="fatload mmc 0:1 ${scriptaddr} /boot.scr; source ${scriptaddr}"
 CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_USE_SECTOR=y
 CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_SECTOR=0x400
diff --git a/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-ostree-scr-fit/apalis-imx8/boot.cmd b/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-ostree-scr-fit/apalis-imx8/boot.cmd
index 98f2e539..57e30bac 100644
--- a/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-ostree-scr-fit/apalis-imx8/boot.cmd
+++ b/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-ostree-scr-fit/apalis-imx8/boot.cmd
@@ -11,9 +11,6 @@ setenv rootpart 2
 setenv hdmi_image hdmitxfw.bin
 setenv m4_0_image core0_m4_image.bin
 setenv m4_1_image core1_m4_image.bin
-setenv ramdisk_addr_r 0x8a000000
-# enable relocation of ramdisk
-setenv initrd_high

 # Boot image files
 setenv fit_addr ${ramdisk_addr_r}
@@ -40,8 +37,8 @@ setenv bootloader2_s_image ${bootloader2_image}
 setenv update_image_boot0 'echo "${fio_msg} writing ${image_path} ..."; run set_blkcnt && mmc dev ${devnum} 1 && mmc write ${loadaddr} ${start_blk} ${blkcnt}'
 setenv update_image_boot1 'echo "${fio_msg} writing ${image_path} ..."; run set_blkcnt && mmc dev ${devnum} 2 && mmc write ${loadaddr} ${start_blk} ${blkcnt}'

-setenv backup_boot0 'echo "${fio_msg} backing up primary boot image set ..."; run set_blkcnt && mmc dev ${devnum} 1 && mmc read ${loadaddr} 0x0 0x3FFE && mmc dev ${devnu>
-setenv restore_boot0 'echo "${fio_msg} restore primary boot image set ..."; run set_blkcnt && mmc dev ${devnum} 2 && mmc read ${loadaddr} 0x0 0x3FFE && mmc dev ${devnum}>
+setenv backup_boot0 'echo "${fio_msg} backing up primary boot image set ..."; run set_blkcnt && mmc dev ${devnum} 1 && mmc read ${loadaddr} 0x0 0x3FFE && mmc dev ${devnum} 2 && mmc write ${loadaddr} 0x0 0x3FFE'
+setenv restore_boot0 'echo "${fio_msg} restore primary boot image set ..."; run set_blkcnt && mmc dev ${devnum} 2 && mmc read ${loadaddr} 0x0 0x3FFE && mmc dev ${devnum} 1 && mmc write ${loadaddr} 0x0 0x3FFE'

 setenv update_primary_image "run update_image_boot0"
 setenv update_primary_image2 "run update_image_boot0"
```

ricardosalveti reviewed (approved):
LGTM, will let @igoropaniuk do a final review.

igoropaniuk commented on meta-lmp-bsp/recipes-support/mfgtool-files/mfgtool-files/imx8qm-mek/full_image.uuu.in:
```diff
@@ -21,7 +21,7 @@ SDPV: jump
 
 FB: ucmd setenv fastboot_dev mmc
 FB: ucmd setenv mmcdev 0
-FB: ucmd mmc dev ${mmcdev} 1; mmc erase 0 0x3FFE
```
We don't need to do that for iMX8QM-MEK, as Toradex config block is not used here

igoropaniuk replied:
same for `imx8qm-mek/bootloader.uuu.in`

ricardosalveti replied:
Indeed, nice catch.

igoropaniuk replied:
@ricardosalveti that was @MrCry0 who initially noticed that :)

MrCry0 replied:
> @ricardosalveti that was @MrCry0 who initially noticed that (we even discussed that in Slack before) :)

Thanks to @igoropaniuk, he noted there were still no my comments :)

It was funny, I wrote comments, but as a part of a review and forgot to submit it :-D

mike-scott replied:
Dropped this patch.  Good catch.

igoropaniuk reviewed (commented):
LGTM, only one minor comment

igoropaniuk commented:
@mike-scott we should also drop these two config options:
```
CONFIG_BOOTCOUNT_LIMIT=y
CONFIG_BOOTCOUNT_ENV=y
```

as it's handled now in boot.cmd instead (and to avoid collisions in logic)

did that already for imx8qm-mek https://github.com/foundriesio/meta-lmp/pull/459/files

mike-scott commented on meta-lmp-bsp/recipes-support/mfgtool-files/mfgtool-files/imx8qm-mek/bootloader.uuu.in:
```diff
@@ -21,7 +21,7 @@ SDPV: jump
 FB: ucmd setenv fastboot_dev mmc
 FB: ucmd setenv mmcdev 0
-FB: ucmd mmc dev ${mmcdev} 1; mmc erase 0 0x3FFE
+FB: ucmd mmc dev ${mmcdev} 1; mmc erase 0 0x3C00
```
@MrCry0 Question is should we just erase `0x4000` here?  Because as is, this command is not working as intended.

mike-scott replied:
Dropped this patch.

mike-scott commented on meta-lmp-bsp/recipes-support/mfgtool-files/mfgtool-files/apalis-imx8/bootloader.uuu.in:
```diff
@@ -21,7 +21,7 @@ SDPV: jump
 FB: ucmd setenv fastboot_dev mmc
 FB: ucmd setenv mmcdev 0
-FB: ucmd mmc dev ${mmcdev} 1; mmc erase 0 0x3FFE
+FB: ucmd mmc dev ${mmcdev} 1; mmc erase 0 0x3C00
```
Done

mike-scott commented:
> @mike-scott we should also drop these two config options:
> 
> ```
> CONFIG_BOOTCOUNT_LIMIT=y
> CONFIG_BOOTCOUNT_ENV=y
> ```
> 
> as it's handled now in boot.cmd instead (and to avoid collisions in logic)
> 
> did that already for imx8qm-mek https://github.com/foundriesio/meta-lmp/pull/459/files

Done.

mike-scott commented:
- Rebased on current `master`
- Addressed comments
- Added `apalis-imx8` changes to align with https://github.com/foundriesio/meta-lmp/pull/459

Pertninent diff from last version:
```
diff --git a/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-fio/apalis-imx8/lmp.cfg b/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-fio/apalis-imx8/lmp.cfg
index d898633d..91322d40 100644
--- a/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-fio/apalis-imx8/lmp.cfg
+++ b/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-fio/apalis-imx8/lmp.cfg
@@ -8,6 +8,7 @@ CONFIG_SPL_FIT_SIGNATURE=y
 CONFIG_SPL_FIT_SIGNATURE_STRICT=y
 CONFIG_FIT_VERBOSE=y
 CONFIG_SPL_LOAD_FIT=y
+CONFIG_PSCI_BOARD_REBOOT=y
 # CONFIG_SPL_FIT_IMAGE_TINY is not set
 # CONFIG_SPL_SATA_SUPPORT is not set
 # CONFIG_SPL_LOAD_IMX_CONTAINER is not set
@@ -27,8 +28,6 @@ CONFIG_ENV_SIZE=0x4000
 CONFIG_CMD_IMPORTENV=y
 CONFIG_CMD_EDITENV=y
 CONFIG_CMD_SAVEENV=y
-CONFIG_BOOTCOUNT_LIMIT=y
-CONFIG_BOOTCOUNT_ENV=y
 # CONFIG_DISTRO_DEFAULTS is not set
 CONFIG_CMD_EXT4=y
 CONFIG_CMD_FAT=y
diff --git a/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-ostree-scr-fit/apalis-imx8/boot.cmd b/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-ostree-scr-fit/apalis-imx8/boot.cmd
index 57e30bac..951dd269 100644
--- a/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-ostree-scr-fit/apalis-imx8/boot.cmd
+++ b/meta-lmp-bsp/recipes-bsp/u-boot/u-boot-ostree-scr-fit/apalis-imx8/boot.cmd
@@ -43,5 +43,7 @@ setenv restore_boot0 'echo "${fio_msg} restore primary boot image set ..."; run
 setenv update_primary_image "run update_image_boot0"
 setenv update_primary_image2 "run update_image_boot0"

+setenv do_reboot "reboot"
+
 @@INCLUDE_COMMON_IMX@@
 @@INCLUDE_COMMON_ALTERNATIVE@@
```

mike-scott commented on meta-lmp-bsp/recipes-support/mfgtool-files/mfgtool-files/imx8qm-mek/full_image.uuu.in:
```diff
@@ -21,7 +21,7 @@ SDPV: jump
 FB: ucmd setenv fastboot_dev mmc
 FB: ucmd setenv mmcdev 0
-FB: ucmd mmc dev ${mmcdev} 1; mmc erase 0 0x3FFE
+FB: ucmd mmc dev ${mmcdev} 1; mmc erase 0 0x3C00
```
Dropped this patch.

mike-scott commented:
- Dropped erase block change for `imx8qm-mek`

MrCry0 reviewed (approved):
LGTM

mike-scott commented:
Fixed the disable of `bootcount` handling in u-boot:
```
-CONFIG_BOOTCOUNT_LIMIT=y
-CONFIG_BOOTCOUNT_ENV=y
+# CONFIG_BOOTCOUNT_LIMIT is not set
```

igoropaniuk reviewed (approved):
LGTM
